### PR TITLE
Refactor and Simplify Codebase for Harbor Satellite

### PR DESCRIPTION
**Pull request by bupd**
## Overview:
The current Harbor Satellite codebase is characterized by complex logic and a lack of clear flow, making it challenging for  future contributors to understand its functionality and make necessary modifications. The intricacies of the existing code structure hinder productivity and increase the likelihood of introducing errors during updates or enhancements.

### Changes Made:

- **Modularization:** 
  Refactored large, monolithic functions into smaller, well-defined functions to improve clarity and reusability.

- **Improved Naming Conventions:** 
  Updated variable and function names to be more descriptive and reflective of their roles in the code.

- **Simplified Logic:** 
  Rewrote complex conditional statements and loops to streamline the control flow.

- **Enhanced Documentation:** 
  Added detailed inline comments and created a separate document outlining the new code flow and logic.

### Expected Outcome:

- A codebase that is significantly more readable and maintainable.
- Enhanced developer productivity and confidence in making future enhancements and modifications.
- Reduced complexity and ambiguity, resulting in fewer errors and easier onboarding for new contributors.

Fixes: #29 

## Summary by CodeRabbit

- **New Features**
  - Updated registry configuration to use remote URLs for better connectivity and reliability.
  - Enhanced error handling and logging for more robust application performance.

- **Bug Fixes**
  - Improved error handling in the `Satellite` component to prevent failures during replication.

- **Refactor**
  - Renamed methods for clarity and consistency across the application.
  - Modularized input processing functions for better code maintainability.

- **Tests**
  - Adjusted test cases to reflect updated image operations and handling of standard output.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -108,42 +99,22 @@ func run() error {
 	}
 
 	input := viper.GetString("url_or_file")
+	// Attempt to parse the input as a URL
 	parsedURL, err := url.Parse(input)
+	// If parsing as URL fails or no scheme detected, treat it as a file path
```
[the remark this answers is not in the document]
Comment (bupd) [reply main.go:104]: We handle it differently in this case, We are checking if it's a url or file. And the err is omitted.

File: main.go [hunk 2]
```diff
@@ -165,3 +136,54 @@ func run() error {
...
+		fmt.Println("Error getting current directory:", err)
+		return err
+	}
+	absPath := filepath.Join(dir, input)
+	if _, err := os.Stat(absPath); os.IsNotExist(err) {
+		fmt.Println("No URL or file found. Please check the configuration.")
+		return fmt.Errorf("file not found")
+	}
+	fmt.Println("Input is a valid file path.")
+	fetcher = store.FileImageListFetcher(input)
```
[the remark this answers is not in the document]
Comment (bupd) [reply main.go:157]: fixed

File: internal/store/in-memory-store.go [hunk 3]
```diff
@@ -170,13 +203,13 @@ func (s *inMemoryStore) Add(ctx context.Context, digest string, image string) er
 	}
 }
 
-func (s *inMemoryStore) AddImage(ctx context.Context, image string) error {
+func (s *inMemoryStore) AddImage(ctx context.Context, image string) {
 	// Add the image to the store
 	s.images[image] = ""
 	fmt.Printf("Added image: %s\n", image)
-	return nil
 }
```
[the remark this answers is not in the document]
Comment (bupd) [reply internal/store/in-memory-store.go:210]: not a good one

File: internal/store/in-memory-store.go [hunk 4]
```diff
@@ -190,11 +223,10 @@ func (s *inMemoryStore) Remove(ctx context.Context, digest string, image string)
 	}
 }
 
-func (s *inMemoryStore) RemoveImage(ctx context.Context, image string) error {
-	// Remove the image from the store
+// Remove the image from the store
+func (s *inMemoryStore) RemoveImage(ctx context.Context, image string) {
 	delete(s.images, image)
 	fmt.Printf("Removed image: %s\n", image)
```
[the remark this answers is not in the document]
Comment (bupd) [reply internal/store/in-memory-store.go:229]: done

File: internal/store/in-memory-store.go [hunk 5]
```diff
@@ -170,13 +206,13 @@ func (s *inMemoryStore) Add(ctx context.Context, digest string, image string) er
 	}
 }
 
-func (s *inMemoryStore) AddImage(ctx context.Context, image string) error {
+func (s *inMemoryStore) AddImage(ctx context.Context, image string) {
 	// Add the image to the store
 	s.images[image] = ""
 	fmt.Printf("Added image: %s\n", image)
-	return nil
 }
```
[the remark this answers is not in the document]
Comment (bupd) [reply internal/store/in-memory-store.go:213]: not a great suggestion, considering running on edge location.

**Comment by bupd** [comment]
@Vad1mo ready for review!

**Review by Vad1mo — CHANGES_REQUESTED** [review]

File: internal/replicate/replicate.go [hunk 6]
```diff
@@ -13,80 +13,75 @@ import (
...
 )
 
+// Replicator interface for image replication and deletion.
 type Replicator interface {
-	// Replicate copies images from the source registry to the local registry.
 	Replicate(ctx context.Context, image string) error
 	DeleteExtraImages(ctx context.Context, imgs []store.Image) error
 }
 
+// BasicReplicator implements the Replicator interface.
```
[the remark this answers is not in the document]
Comment (bupd) [reply internal/replicate/replicate.go]: Thanks for pointing out!, Fixed it.

**Comment by bupd** [comment]
Ready to merge!

File: test/e2e/satellite_test.go [hunk 7]
```diff
@@ -155,16 +151,14 @@ func buildSatellite(
...
 		WithExec([]string{"go", "build", "-o", appBinary, sourceFile}).
 		WithExposedPort(9090).
 		WithExec([]string{"go", "run", "./test/e2e/test.go"})
 
 	assert.NoError(t, err, "Test failed in buildSatellite")
 
-	stdOut, err := container.Stdout(ctx)
-	assert.NoError(t, err, "Failed to get stdOut in Satellite")
-
+	stdOut, _ := container.Stdout(ctx)
```
[the remark this answers is not in the document]
Comment (bupd) [reply test/e2e/satellite_test.go:161]: the only reason for accessing the stdout of container to make just so dagger not cache it.

**Pull request merged**